File: backtracking/39-Combination-Sum/neet.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
  const res = [];
  const cur = [];

  const dfs = (i, total) => {
    if (total === target) {
      res.push([...cur]);
      return;
    }
    if (i >= candidates.length || total > target) {
      return;
    }
    cur.push(candidates[i]);
    dfs(i, total + candidates[i]);
    cur.pop();
    dfs(i + 1, total);
  };

  dfs(0, 0);

  return res;
};

File: backtracking/39-Combination-Sum/neet.py
class Solution:
    def combinationSum(self, candidates, target: int):
        res, cur = [], []

        def dfs(i, total):
            if total == target:
                res.append(cur.copy())
                return
            if i >= len(candidates) or total > target:
                return
            
            cur.append(candidates[i])
            dfs(i, total + candidates[i])
            cur.pop()
            dfs(i + 1, total)

        dfs(0, 0)
        return res    